accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,189 @@
+# Trusted Publishers, using OpenID Connect for NuGet push (technical)
+
+- Author: Joel Verhagen ([@joelverhagen](https://github.com/joelverhagen) on GitHub)
+- Issue: [NuGet/NuGetGallery#9332](https://github.com/NuGet/NuGetGallery/issues/9332)
+
+This is the technical description of the experience described in [Trusted Publishers, using OpenID Connect for NuGet
+push](trusted-publishers-oidc-for-nuget-push.md). The content is split to clearly separate the functional/UX description
+(the other doc) from the technical description (this doc). This document acts as a supporting document to the other so
+that other one should be read first.
+
+Much of the technical explanation is described in [Trusted Publishers for All Package
+Repositories](https://repos.openssf.org/trusted-publishers-for-all-package-repositories). I will expand on certain
+technical details which are particularly interesting or specific to NuGet.
+
+## Validation of the GitHub OIDC token
+
+The most critical step in this design is understanding how NuGet.org will validate that an incoming OIDC token is
+acceptable to trade for a short-lived NuGet API key.
+
+The following checks are made:
+
+- The token is issued by a known Trusted Publisher (only GitHub Actions at this time)
+- The token is validated per all JWT rules, such as but not limited to:
+  - valid signature via JWKS
+  - validate duration (`nbf` and `exp` claims)
+  - only used once, via `jti`
+  - valid `aud` claim, being `nuget`
+
+The
+[Microsoft.IdentityModel.Protocols.OpenIdConnect](https://www.nuget.org/packages/Microsoft.IdentityModel.Protocols.OpenIdConnect)
+package can help us properly validate JWTs in our NuGet/NuGetGallery ASP.NET application.
+
+In addition to the general JWT checks, specific checks are made for each Trusted Publisher. For GitHub Actions, the
+following checks will be made:
+
+- `sub` claim matches `{repo owner}/{repo name}:.*` (case insensitive)
```

Comment:
It's a bit difficult for me to understand what's variable and what's constant.  For example, this might be reworded to:

`sub` claim matches the regular expression pattern `{repo owner}/{repo name}:.*` where `{repo owner}` is the repository owner's name and `{repo name}` is the repository's name.  Then, I would understand `dtivel/myrepo:somepath` as a matching pattern.

(I inferred that `*` is something of a wildcard or maybe `.*` is a regular expression pattern, but I can't be certain from the spec.)

What would an actual `sub` look like?

Reply:
I expressed it as a regex. You're right, that's confusing. I'll clarify that and add a sample.

A branch `sub` would look like this: `repo:octo-org/octo-repo:ref:refs/heads/demo-branch`.